// emberfire/addon/services/firebase-app.js
import Service from '@ember/service';
import { get, set } from '@ember/object';
import { inject as service } from '@ember/service';
import RSVP from 'rsvp';
const { resolve } = RSVP;
const getApp = (service) => {
    const firebase = get(service, 'firebase');
    const name = get(service, 'name');
    return firebase.app(name);
};
export default class FirebaseAppService extends Service.extend({
    name: undefined,
    firebase: service('firebase')
}) {
    constructor() {
        super(...arguments);
        this.delete = () => getApp(this).delete();
        this.auth = () => resolve(import('firebase/auth')).then(() => getApp(this).auth());
        this.firestore = () => resolve(import('firebase/firestore')).then(() => getApp(this).firestore());
        this.messaging = () => resolve(import('firebase/messaging')).then(() => getApp(this).messaging());
        this.database = (url) => resolve(import('firebase/database')).then(() => getApp(this).database(url));
        this.functions = (region) => resolve(import('firebase/functions')).then(() => getApp(this).functions(region));
        // TODO: drop the <any> once firebase-js-sdk #1792 and #1812 and patched in, they were missing typings
        this.performance = () => resolve(import('firebase/performance')).then(() => getApp(this).performance());
        this.storage = (url) => resolve(import('firebase/storage')).then(() => getApp(this).storage(url));
    }
    init() {
        this._super(...arguments);
        const app = getApp(this);
        set(this, 'options', app.options);
    }
}
